nix-serve: fix NIX_SECRET_KEY_FILE

Artturin opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Artturin committed:
nix-serve: fix NIX_SECRET_KEY_FILE

pennae commented on nixos/modules/services/networking/nix-serve.nix:
```diff
@@ -61,16 +61,19 @@ in
-        ExecStart = "${pkgs.nix-serve}/bin/nix-serve " +
+        # does not work due to the cred var not being substituted
+        # environment.NIX_SECRET_KEY_FILE = "$CREDENTIALS_DIRECTORY/NIX_SECRET_KEY_FILE";
+        ExecStart = "NIX_SECRET_KEY_FILE=$CREDENTIALS_DIRECTORY/NIX_SECRET_KEY_FILE " +
+          "${pkgs.nix-serve}/bin/nix-serve " +
           "--listen ${cfg.bindAddress}:${toString cfg.port} ${cfg.extraParams}";
```
```suggestion
        script = ''
          export NIX_SECRET_KEY_FILE="$CREDENTIALS_DIRECTORY/NIX_SECRET_KEY_FILE"
          exec ${pkgs.nix-serve}/bin/nix-serve \
            --listen ${escapeShellArg cfg.bindAddress}:${toString cfg.port} ${cfg.extraParams}
        '';
```

alexbakker commented on nixos/modules/services/networking/nix-serve.nix:
```diff
@@ -61,16 +61,19 @@ in
+
+      script = ''
+        export NIX_SECRET_KEY_FILE="$CREDENTIALS_DIRECTORY/NIX_SECRET_KEY_FILE"
+        ${pkgs.nix-serve}/bin/nix-serve --listen ${cfg.bindAddress}:${toString cfg.port} ${cfg.extraParams}
```
Should probably prepend ``exec`` here.

alexbakker commented:
I left one nitpick comment.

Other than that: I tested this on 21.11 and it works well.

Mic92 commented:
@ofborg test nix-serve

Mic92 commented on nixos/modules/services/networking/nix-serve.nix:
```diff
@@ -61,16 +59,19 @@ in
         User = "nix-serve";
         Group = "nix-serve";
         DynamicUser = true;
+        LoadCredential = "NIX_SECRET_KEY_FILE:${cfg.secretKeyFile}";
```
```suggestion
        LoadCredential = lib.optionalString (cfg.secretKeyFile != null) "NIX_SECRET_KEY_FILE:${cfg.secretKeyFile}";
```

Artturin replied:
thanks

Mic92 commented on nixos/modules/services/networking/nix-serve.nix:
```diff
@@ -61,16 +59,19 @@ in
-      environment.NIX_SECRET_KEY_FILE = cfg.secretKeyFile;
+
+      script = ''
+        export NIX_SECRET_KEY_FILE="$CREDENTIALS_DIRECTORY/NIX_SECRET_KEY_FILE"
```
This also might have to be wrapped in an optional string than if we have no secret file.

Artturin commented:
the test fails with `vm-test-run-nix-ssh-serve> client # error: experimental Nix feature 'nix-command' is disabled; use '--extra-experimental-features nix-command' to override`

Mic92 commented:
> the test fails with `vm-test-run-nix-ssh-serve> client # error: experimental Nix feature 'nix-command' is disabled; use '--extra-experimental-features nix-command' to override`

I cannot reproduce this error locally, can you?

Artturin commented:
> > the test fails with `vm-test-run-nix-ssh-serve> client # error: experimental Nix feature 'nix-command' is disabled; use '--extra-experimental-features nix-command' to override`
> 
> I cannot reproduce this error locally, can you?

yes at https://github.com/nixos/nixpkgs/commit/a78dd785b29bbbcad5490282308b9dbdc244a0f3

Mic92 commented:
It's not even clear to me why there is `nix copy` used.

Artturin commented:
there's 3 nix commands run in nix-serve https://github.com/edolstra/nix-serve/blob/fefa88d82ead34d1ad97cc68cccae9412accfc1f/nix-serve.psgi

2 `nix dump-path`
1 `nix-log`

so we should pin it to `nix_2_3` for now 
also there's https://github.com/edolstra/nix-serve/issues/28

Artturin commented:
the nix-serve test is defined wrong in all-tests.nix

```
  nix-serve = handleTest ./nix-ssh-serve.nix {};
  nix-ssh-serve = handleTest ./nix-ssh-serve.nix {};
```
nix-ssh-serve has nix copy

Artturin commented:
you forgot to change the path after you pasted in ac7b8724b59974c0d74f2feacc4a2a787a5cf122

Artturin committed:
nixos/tests: fix nix-serve path

nixos/tests: rename nix-ssh-serve to nix-serve-ssh

nixos/tests/nix-serve-ssh: add --experimental-features

nixos-serve: add nix-serve-ssh to passthru.tests

Artturin commented:
nevermind no need to pin nix https://github.com/NixOS/nix/pull/5634 this pr fixed the issue
